mm2src/mm2_net/src/wasm/body_stream.rs

```diff
@@ -0,0 +1,400 @@
+    }
+}
+
+/// Handles the HTTP response body, decoding data, and extracting trailers
+#[pin_project]
+pub struct ResponseBody {
+    #[pin]
+    body_stream: BodyStream,
+    buf: EncodedBytes,
+    incomplete_data: BytesMut,
+    data: Option<BytesMut>,
+    trailer: Option<HeaderMap>,
+    state: ReadState,
+    finished_stream: bool,
+}
+
+impl ResponseBody {
+    /// Creates a new `ResponseBody` based on a ReadableStream and content type.
+    pub(crate) fn new(body_stream: ReadableStream, content_type: &str) -> Result<Self, PostGrpcWebErr> {
+        let body_stream = wasm_streams::ReadableStream::from_raw(body_stream.unchecked_into()).into_stream();
+
+        Ok(Self {
+            body_stream: BodyStream::new(body_stream),
+            buf: EncodedBytes::new(content_type)?,
+            incomplete_data: BytesMut::new(),
+            data: None,
+            trailer: None,
+            state: ReadState::CompressionFlag,
+            finished_stream: false,
+        })
+    }
+
+    fn read_stream(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), PostGrpcWebErr>> {
+        if self.finished_stream {
+            return Poll::Ready(Ok(()));
+        }
+
+        let this = self.project();
+
+        match ready!(this.body_stream.poll_data(cx)) {
+            Some(Ok(data)) => {
+                if let Err(e) = this.buf.append(data) {
+                    return Poll::Ready(Err(e));
+                }
+
+                Poll::Ready(Ok(()))
+            },
+            Some(Err(e)) => Poll::Ready(Err(e)),
+            None => {
+                *this.finished_stream = true;
+                Poll::Ready(Ok(()))
+            },
+        }
+    }
+
+    fn step(self: Pin<&mut Self>) -> Result<(), PostGrpcWebErr> {
+        let this = self.project();
+
+        loop {
+            match this.state {
+                ReadState::CompressionFlag => {
+                    if this.buf.is_empty() {
+                        // Can't read compression flag right now
+                        return Ok(());
+                    } else {
+                        let compression_flag = this.buf.take(1);
+
+                        if compression_flag[0] & TRAILER_BIT == 0 {
+                            this.incomplete_data.unsplit(compression_flag);
+                            *this.state = ReadState::DataLength;
+                        } else {
+                            *this.state = ReadState::TrailerLength;
+                        }
+                    }
+                },
+                ReadState::DataLength => {
+                    if this.buf.len() < 4 {
+                        // Can't read data length right now
+                        return Ok(());
+                    } else {
+                        let data_length_bytes = this.buf.take(4);
+                        let data_length = BigEndian::read_u32(data_length_bytes.as_ref());
+
+                        this.incomplete_data.unsplit(data_length_bytes);
```

Comment:
According to docs (https://docs.rs/bytes/1.4.0/bytes/struct.BytesMut.html#method.unsplit), `unsplit` is usable for bytes that were previously `split_off`, which is not the case here. Please consider alternative solution.

Reply:
https://docs.rs/bytes/1.4.0/bytes/struct.BytesMut.html#method.extend_from_slice it is then..thank you